opnode/l2/api.go

```diff
@@ -111,6 +115,48 @@ func (payload *ExecutionPayload) ID() eth.BlockID {
 	return eth.BlockID{Hash: payload.BlockHash, Number: uint64(payload.BlockNumber)}
 }
 
+func (payload *ExecutionPayload) ParentID() eth.BlockID {
+	n := uint64(payload.BlockNumber)
+	if n > 0 {
+		n -= 1
+	}
+	return eth.BlockID{Hash: payload.ParentHash, Number: n}
+}
+
+type rawTransactions []Data
+
+func (s rawTransactions) Len() int { return len(s) }
+func (s rawTransactions) EncodeIndex(i int, w *bytes.Buffer) {
+	w.Write(s[i])
+}
+
+// CheckBlockHash recomputes the block hash and returns if the embedded block hash matches.
+func (payload *ExecutionPayload) CheckBlockHash() (actual common.Hash, ok bool) {
+	hasher := trie.NewStackTrie(nil)
+	txHash := types.DeriveSha(rawTransactions(payload.Transactions), hasher)
+
+	header := types.Header{
+		ParentHash:  payload.ParentHash,
+		UncleHash:   types.EmptyUncleHash,
+		Coinbase:    payload.FeeRecipient,
+		Root:        common.Hash(payload.StateRoot),
+		TxHash:      txHash,
+		ReceiptHash: common.Hash(payload.ReceiptsRoot),
+		Bloom:       types.Bloom(payload.LogsBloom),
+		Difficulty:  big.NewInt(0), // zeroed, proof-of-work legacy
```

Comment:
Thoughts on using `common.Big0` to save an alloc here?
https://github.com/ethereum/go-ethereum/blob/86d547707965685cef732aa28c15e6811ea98408/common/big.go#L26